alerts/config.go

```diff
@@ -5,116 +5,179 @@ package alerts
 }
 
 // Base describes the minimum specifics of a desired alert.
 type Base struct {
 	// Enabled controls whether notifications should be sent for this alert.
-	Enabled bool
+	Enabled bool `json:"enabled"`
 	// Repeat is measured in minutes.
-	Repeat DurationMinutes `json:"repeat"`
+	//
+	// A value of 0 (the default) disables repeat notifications.
+	Repeat DurationMinutes `json:"repeat,omitempty"`
 }
 
 func (b Base) Validate(validator structure.Validator) {
 	validator.Bool("enabled", &b.Enabled)
 	dur := b.Repeat.Duration()
-	validator.Duration("repeat", &dur).GreaterThan(0 * time.Minute)
+	validator.Duration("repeat", &dur).Using(validateRepeat)
 }
 
-// DelayMixin adds a configurable delay.
-type DelayMixin struct {
-	// Delay is measured in minutes.
-	Delay DurationMinutes `json:"delay,omitempty"`
-}
+const (
+	// RepeatMin is the minimum duration for a repeat setting (if not 0).
+	RepeatMin = 15 * time.Minute
+	// RepeatMax is the maximum duration for a repeat setting.
+	RepeatMax = 4 * time.Hour
+)
 
-func (d DelayMixin) Validate(validator structure.Validator) {
-	dur := d.Delay.Duration()
-	validator.Duration("delay", &dur).GreaterThan(0 * time.Minute)
+func validateRepeat(value time.Duration, errorReporter structure.ErrorReporter) {
+	if value == 0 {
+		return
+	}
+	if value < RepeatMin {
+		errorReporter.ReportError(validator.ErrorValueNotGreaterThanOrEqualTo(value, RepeatMin))
+	}
+	if value > RepeatMax {
+		errorReporter.ReportError(validator.ErrorValueNotLessThanOrEqualTo(value, RepeatMax))
+	}
 }
 
-// ThresholdMixin adds a configurable threshold.
-type ThresholdMixin struct {
+// UrgentLowAlert extends Base with a threshold.
+type UrgentLowAlert struct {
+	Base `bson:",inline"`
 	// Threshold is compared the current value to determine if an alert should
 	// be triggered.
 	Threshold `json:"threshold"`
 }
 
-func (t ThresholdMixin) Validate(validator structure.Validator) {
-	t.Threshold.Validate(validator)
+func (d UrgentLowAlert) Validate(validator structure.Validator) {
+	d.Base.Validate(validator)
+	d.Threshold.Validate(validator)
+	validateThresholdInNativeUnits(validator, d.Threshold, 40, 55, glucose.MgdL)
 }
 
-// WithThreshold extends Base with ThresholdMixin.
-type WithThreshold struct {
-	Base           `bson:",inline"`
-	ThresholdMixin `bson:",inline"`
+// validateThresholdInNativeUnits formats validation errors in the same units
+// submitted by the user.
+//
+// This provides easier to understand validation error messages.
+func validateThresholdInNativeUnits(validator structure.Validator, threshold Threshold, min, max float64, units string) {
+	nativeMin := glucose.Convert(min, glucose.MgdL, threshold.Units)
+	nativeMax := glucose.Convert(max, glucose.MgdL, threshold.Units)
+	validator.Float64("threshold.value", &threshold.Value).InRange(nativeMin, nativeMax)
+}
+
+// NotLoopingAlert extends Base with a delay.
+type NotLoopingAlert struct {
+	Base  `bson:",inline"`
+	Delay DurationMinutes `json:"delay,omitempty"`
```

Comment:
I don't have context into any previous discussions, what are we trying to accomplish with a custom `DurationMinutes` type here, rather than just an `*int`?